Source/ARTPushActivationStateMachine.m

```diff
@@ -52,6 +57,14 @@ - (instancetype)init:(ARTRestInternal *)rest {
         if (!_pendingEvents) {
             _pendingEvents = [NSMutableArray array];
         }
+
+        // Due to bug #966, old versions of the library might have led us to an illegal
+        // persisted state: we have a deviceToken, but the persisted push state is WaitingForPushDeviceDetails.
+        // So we need to re-emit the GotPushDeviceDetails event that led us there.
+        if ([_current isKindOfClass:[ARTPushActivationStateWaitingForPushDeviceDetails class]] && rest.device_nosync.deviceToken != nil) {
+            [rest.logger debug:@"ARTPush: re-emitting stored device details for stuck state machine"];
+            [self handleEvent:[ARTPushActivationEventGotPushDeviceDetails new]];
```

Comment:
Knowing very little about this code but aware that we want to get this out soon, please excuse my possibly naive question... is this not what `_pendingEvents` (defined above) could be used for? My curiosity spiked by how this seems to be synchronous event handling from within a initialiser - which, just, smells a little wrong.

Reply:
@QuintinWillison That's for events that arrived when the state machine was at a state that didn't have a transition defined for them. Those events are enqueued and, when the machine transitions to a new state, they're reprocessed, in case this new state does define a transition.

That's not the case here. This is a new event, not an old event that didn't get processed.

Reply:
> this seems to be synchronous event handling from within a initialiser - which, just, smells a little wrong.

Well, it _is_ initialization code. It's correcting the wrong state that was persisted. No code should see that state.

Reply:
Great, thanks for explaining. 😎